### DISCUSSION TITLE
批量拉取数量不对

### DISCUSSION THREAD

**Discussion by 1025044866xxy** [apache/rocketmq#7596] [Q&A] [answered]
我使用的是DefaultMQPushConsumer api，设置了pullBatchSize和consumeMessageBatchMaxSize都为512，同时修改了broker相关参数,maxTransferCountOnMessageInMemory,maxTransferBytesOnMessageInMemory,maxTransferBytesOnMessageInDisk,maxTransferCountOnMessageInDisk。为什么在消息堆积的时候，我拉取到的消息数还是很少？不是我期望的拉满呢？比如说拉到512条。

**Comment by l0ady** [comment, marked as answer]
看代码这个不仅和消息的条数有关系，和每条消息的大小也有关系；

**Reply by l0ady** [reply]
maxTransferCountOnMessageInMemory,maxTransferBytesOnMessageInMemory,maxTransferBytesOnMessageInDisk,maxTransferCountOnMessageInDisk这些参数分别设置的值是多少？

**Reply by 1025044866xxy** [reply]
maxTransferCountOnMessageInMemory 8192

maxTransferCountOnMessageInDisk 1024

maxTransferBytesOnMessageInMemory 33554432

maxTransferBytesOnMessageInDisk 8388608

**Reply by 1025044866xxy** [reply]
我也看过代码，当读取的消息数或者对应的字节数，超过设置的上限时，就会提前返回。这个对应的bytes或者count，是针对全部topic一起的值吗？ 因为就字节数来说，是内存里面类似于有缓冲池的值总和吗？

**Reply by 1025044866xxy** [reply]
因为就设置来看，我的上限很高，如果是单个toipc，产生了堆积。那么一次能拉取的消息数应该很多才对，但是还是很少，不会达到上限。

**Comment by letian-tang** [comment]
我也遇到同样的问题，后来解决没有啊？

**Reply by 1025044866xxy** [reply]
上面的设置是生效的，我发现部分机器会拉满数据，部分不可以，还在找原因。

**Reply by 1025044866xxy** [reply]
看了下源码，猜测还有一个可能性就是，consumeQueue索引生成的速度太慢了，导致消息不可见。可以改成并行生成试试。https://github.com/apache/rocketmq/issues/5184
